src/Microsoft.AspNetCore.WebSockets/WebSocketMiddleware.cs

```diff
@@ -42,7 +45,24 @@ public Task Invoke(HttpContext context)
             var upgradeFeature = context.Features.Get<IHttpUpgradeFeature>();
             if (upgradeFeature != null && context.Features.Get<IHttpWebSocketFeature>() == null)
             {
-                context.Features.Set<IHttpWebSocketFeature>(new UpgradeHandshake(context, upgradeFeature, _options));
+                var webSocketFeature = new UpgradeHandshake(context, upgradeFeature, _options);
+                context.Features.Set<IHttpWebSocketFeature>(webSocketFeature);
+
+                if (_options.AllowedOrigins.Count > 0)
+                {
+                    // Check for Origin header
+                    var originHeader = context.Request.Headers[HeaderNames.Origin];
+
+                    if (!StringValues.IsNullOrEmpty(originHeader) && webSocketFeature.IsWebSocketRequest)
+                    {
+                        // Check allowed origins to see if request is allowed
+                        if (!_options.AllowedOrigins.Contains(originHeader.ToString(), StringComparer.Ordinal) && !_options.AllowedOrigins.Contains("*", StringComparer.Ordinal))
+                        {
+                            context.Response.StatusCode = StatusCodes.Status403Forbidden;
```

Comment:
Log

Reply:
And again, duplication. Might it not be better to take the pieces from the existing CORS implementation and have an internal or helper class which has this code, so if it has problems it's a fix in a single place?

Reply:
Log: ✔️ 
Reduce Duplication: As long as there's something consistent to extract here and it's relatively straightforward to extract I'm cool with that. Is this representative of the logic in CORS? If so, it doesn't seem complex enough to warrant a whole new helper (given that means a shared-source package as well) to me.